### PR TITLE
audiobookshelf: init at 2.2.15

### PR DESCRIPTION
**Pull request by jvanbruegge**
###### Description of changes

Audiobookshelf is a self-hosted alternative to services like audible

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jvanbruegge** [comment]
I don't understand the error that ofborg throws

**Comment by jvanbruegge** [comment]
I used `buildNpmPackage` for the client now so I don't have to commit a 14k LOC file with the dependencies

**Comment by jvanbruegge** [comment]
But I still don't know what ofborg wants here

File: pkgs/servers/audiobookshelf/default.nix [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      PORT="$port" \
+      HOST="$host" \' > $out/bin/${pname}
+    echo "  exec ${nodejs}/bin/node $out/opt/index.js" >> $out/bin/${pname}
+    chmod +x $out/bin/${pname}
+  '';
+
+  passthru.updateScript = ./generate.sh;
+
+
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/default.nix:114]:
```suggestion

  meta = with lib; {
```

File: pkgs/servers/audiobookshelf/default.nix [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  buildInputs = [ nodejs ffmpeg-full tone util-linux ];
+
+  installPhase = ''
+    mkdir -p $out/opt/client
+    cp -r index.js server package* $out/opt/
+    cp -r ${client}/lib/node_modules/${pname}-client/dist $out/opt/client/dist
+    mkdir $out/bin
+
+    echo '#!${stdenv.shell}
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/default.nix:46]: We should put the wrapper into an extra file if it gets that big.

File: pkgs/servers/audiobookshelf/default.nix [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+  nodeDeps = (import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  }).nodeDependencies;
+
+in stdenv.mkDerivation rec {
+  inherit pname version src;
+
+  buildInputs = [ nodejs ffmpeg-full tone util-linux ];
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/default.nix:38]: ffmpeg-full and tone are only used in the wrapper, so putting them here will be a noop.

File: pkgs/servers/audiobookshelf/default.nix [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-gQuUYS84U48TA8PsNsvyHmWYynmje9umGvCbNXeQrjo=";
+  };
+
+  client = buildNpmPackage {
+    pname = "${pname}-client";
+    inherit version;
+
+    src = runCommand "cd" {} ''
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/default.nix:20]:
```suggestion
    src = runCommand "cp-source" {} ''
```

File: pkgs/servers/audiobookshelf/generate.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -i bash -p nodePackages.node2nix -p prefetch-npm-deps
+
+src="$(nix-build . -A audiobookshelf.src --no-out-link)"
+
+cd "$(dirname "${BASH_SOURCE[0]}")" || exit 1
+
+hash=$(prefetch-npm-deps "$src/client/package-lock.json")
+echo "npmDepsHash = \"$hash\";" > test.txt
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/generate.sh:9]: Is that file even used?

File: pkgs/applications/audio/tone/default.nix [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{ lib, stdenv, fetchFromGitHub, fetchNuGet, linkFarmFromDrvs, buildDotnetModule, ffmpeg-full, msbuild, dotnetCorePackages }:
+
+let
+  nugetSource = linkFarmFromDrvs "nuget-packages" (
+    import ./nuget-deps.nix { inherit fetchNuGet; }
+  );
+
+in buildDotnetModule rec {
+  pname = "tone";
+  baseName = pname;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/tone/default.nix:10]: You are sure baseName is required?

File: pkgs/servers/audiobookshelf/default.nix [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    NODE_OPTIONS = "--openssl-legacy-provider";
+
+    npmBuildScript = "generate";
+    npmDepsHash = "sha256-UFmdTWxkOGO73g45NzB51/d1oGOhfNZtOlKSfbTsxPY=";
+  };
+
+  nodeDeps = (import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  }).nodeDependencies;
```
Comment (SuperSandro2000) [inline pkgs/servers/audiobookshelf/default.nix:33]: I think this line is causing IFD which is forbidden. I think you must commit the package.json.

File: pkgs/servers/audiobookshelf/generate.sh (hunk 5 above)
Comment (jvanbruegge) [reply pkgs/servers/audiobookshelf/generate.sh]: Nope, that was left over from debugging the sed command

File: pkgs/applications/audio/tone/default.nix (hunk 6 above)
Comment (jvanbruegge) [reply pkgs/applications/audio/tone/default.nix]: Copied it from another dotnet derivation, I don't think so

File: pkgs/servers/audiobookshelf/default.nix (hunk 7 above)
Comment (jvanbruegge) [reply pkgs/servers/audiobookshelf/default.nix]: I found another solution in the derivation for jellyfin, let's see if that works

Comment (jvanbruegge) [reply pkgs/servers/audiobookshelf/default.nix]: Looks like it does

**Comment by jvanbruegge** [comment]
Nice, the tests pass now

**Comment by dit7ya** [comment]
Hey @jvanbruegge, have you seen my PR at #201552? I want to update one little thing in the nixos module, but other than that I was happy with. It does not generate node-deps and has a much shorter diff.

**Comment by jvanbruegge** [comment]
No I did not see that, which is weird because I did use the github search to see if it was already being packaged.

I've done the same as you now for the package, even though that means that it cannot have an update-script any more. (There is no option that simulates --only-production in prefetch-npm-deps).

I also updated audiobookshelf to the newest version (2.2.14).

Could you retarget your branch to this PR and add the service and tests in a seperate commit? I want to keep the wrapper from this PR because I want to run it without the nixos module and without a config file later on.

@SuperSandro2000 I have addressed all your comments

**Comment by jvanbruegge** [comment]
@SuperSandro2000 I've updated to the newest release

**Comment by advplyr** [comment]
Hey, heads up the next release coming in a few days is migrating the file-based JSON database to a sqlite database. This branch https://github.com/advplyr/audiobookshelf/tree/sqlite_2 has the working code. 

This is using [node-sqlite3](https://github.com/TryGhost/node-sqlite3). I have tested the Docker images and debian package. I'm happy to discuss this more if you want here or in discord/matrix.

**Pull request merged**